src/dataregistry/registrar_util.py

```diff
@@ -1,30 +1,44 @@
 import os
+import re
+from sqlalchemy import MetaData, Table, Column, text, select
+
 from dataregistry.db_basic import ownertypeenum
 
-__all__ = ['make_version_string', 'parse_version_string']
+__all__ = ['make_version_string', 'parse_version_string', 'calculate_special',
+           'form_dataset_path', 'get_directory_info']
 VERSION_SEPARATOR = '.'
+_nonneg_int_re = "0|[1-9][0-9]*"
+
 def make_version_string(major, minor=0, patch=0, suffix=None):
     version = VERSION_SEPARATOR.join([major, minor, patch])
     if suffix:
         version = VERSION_SEPARATOR.join([version, suffix])
     return version
 
-def parse_version_string(version):
+def parse_version_string(version, with_suffix=False):
     '''
-    Return dict with keys major, minor, patch and (if present) suffix
+    Return dict with keys major, minor, patch and (if present) suffix.
+    with_suffix == False means version string *must not* include suffix
+    with_suffix == True means it *may* have a suffix
+
+    Returns a dict with keys 'major', 'minor', 'patch' and optionally 'suffix'
     '''
     # Perhaps better to do with regular expressions.  Or at least verify
     # that major, minor, patch are integers
-    cmp = version.split(_SEPARATOR, maxsplit=3)
-    d = {'major' : cmp[0]}
-    if len(cmp) > 1:
-        d['minor'] = cmp[1]
+    cmp = version.split(VERSION_SEPARATOR)
+    if not with_suffix:
+        if len(cmp) != 3:
+            raise ValueError('Version string must have 3 components')
     else:
-        d['minor'] = 0
-    if len(cmp) > 2:
-        d['patch'] = cmp[2]
-    else:
-        d['patch'] = 0
+        if len(cmp) < 3 or len(cmp) > 4:
+            raise ValueError('Version string must have 3 or 4 components')
+    for c in cmp[0:3]:
+        if not re.fullmatch(_nonneg_int_re, c):
+            raise ValueError(f'Version component {c} is not non-negative int')
+    d = {'major' : cmp[0]}
```

Comment:
These should be converted to ints

Reply:
I don't see why.  I've modified header comments to make it clear the values returned are strings.